morpho-checkpoint-node/tests/Test/Morpho/QSM.hs

```diff
@@ -335,46 +354,98 @@ prop_4 = noShrinking $
-            ncNodePort = Right $ fromIntegral $ 3000 + 2 * nodeId,
-            ncValidateDb = Right True
-          }
-      configFile = configDir ++ "/config-" ++ show nodeId ++ ".yaml"
   (mockNode, rpcUpstream) <- mockRpcUpstream
 
-  nodeConfig <- getConfiguration cliConfig configFile
+  let toPublicKey =
+        fromRight' . importPublicKey . fromRight' . bytesFromHex . fromJust . normalizeHex
+  let pubKeys =
+        take
+          nodesNum
+          [ toPublicKey "ec33a3689573db2f4db4586bb7089cda045116a21cce20c9a6fe7ccadcf9fb336075b3644ac9f0a20e6d45a9e99db477cc420d050969f2d8bfb7408b2169b167",
+            toPublicKey "3ea6ebf308a16b45bbf3392d1cd226aa6858d2ea6a2402799eeffc576f0187af62ed11677c8d4d01ef03bbf638f49f2fd77e181924a52dd01b86a187af07b724"
+          ]
+
+  let toPrivKey = fromRight' . importPrivateKey . fromRight' . bytesFromHex . fromJust . normalizeHex
+  let privKey =
+        [ toPrivKey "0093e3cf4be871137e4e11b8d94ec397afb1d3b6db4cdfef01780033b7e3c67f06",
+          toPrivKey "00103becb5b909d34b5a57fc629e305c7ce8f00ef2318939e645a3c161b1d99a03"
+        ]
+          !! nodeId
+
+  systemStart <- parseTimeM True defaultTimeLocale "%FT%XZ" "2020-11-22T04:45:09Z" :: IO UTCTime
+
+  let logFile =
+        "db/qsm-tests-"
+          <> T.pack (show testId)
+          <> "/nodedir-"
+          <> T.pack (show nodeId)
+          <> "/node.log"
+
+  tracers <- mkTracers 0 (fileTrace $ T.unpack logFile)
+
+  let env =
+        Env
+          { eNodeId = CoreNodeId $ fromIntegral nodeId,
+            eNumCoreNodes = fromIntegral nodesNum,
+            eNetworkMagic = NetworkMagic 12345,
+            eSystemStart = SystemStart systemStart,
+            eSecurityParameter = SecurityParam 5,
+            eStableLedgerDepth = 3,
+            eTimeslotLength = mkSlotLength $ secondsToNominalDiffTime 1,
+            eSnapshotsOnDisk = 60,
+            eSnapshotInterval = 60,
+            ePoWBlockFetchInterval = 5000000,
+            ePrometheusPort = 13888 + 2 * nodeId,
+            eCheckpointingInterval = 4,
+            eRequiredMajority = fromIntegral requiredMajority,
+            eFedPubKeys = pubKeys,
+            ePrivateKey = privKey,
+            eTracers = tracers,
+            eRpcUpstream = rpcUpstream,
+            eProducers =
+              [ RemoteAddress "127.0.0.1" (Just 3000) 1,
+                RemoteAddress "127.0.0.1" (Just 3002) 1
+              ],
+            eNodeAddress = NodeAddress (NodeHostAddress Nothing) (fromIntegral $ 3000 + 2 * nodeId),
+            eDatabaseDir = nodeDir ++ "/db",
+            eValidateDb = True
+          }
 
-  node <- async $
-    withEnv nodeConfig $ \(ValidatedEnv env) -> do
-      run $
-        ValidatedEnv $
-          env
-            { eRpcUpstream = rpcUpstream
-            }
+  node <- async $ run $ ValidatedEnv env
 
   link node
   return $ NodeHandle nodeId mockNode node
 
+{-# NOINLINE locallock #-}
+locallock :: MVar ()
+locallock = unsafePerformIO $ newMVar ()
```

Comment:
This is very naughty, and the `stdoutTrace` in iohk-monitoring only works well because there's only a single stdout. In our case there are multiple files, meaning a single lock is shared between all the files, which is not ideal.

Instead you can change `fileTrace` to type `FilePath -> IO (Trace IO T.Text)`, which then allows initializing a new `MVar` within it.